pymllm/engine/launch.py

```
@@ -0,0 +1,377 @@
+        # Start all subprocesses
+        for proc, _, name in procs_and_readers:
+            proc.start()
+            self._subprocesses.append(proc)
+            logger.info("Started %s process (pid=%s)", name, proc.pid)
+
+        # Wait for readiness signals
+        for _, reader, name in procs_and_readers:
+            try:
+                msg = reader.recv()
+            except EOFError:
+                raise RuntimeError(f"{name} process died before signalling readiness")
```

Ruff (0.15.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)